Add GitHub stats to the status dashboard

menghif opened the pull request:
<!--
Thanks for sending a pull request!
- if this is your first time, please ensure you have read through our contributor guide: https://github.com/Seneca-CDOT/telescope/blob/master/docs/CONTRIBUTING.md
-->

## Issue This PR Addresses
Fixes #2417 
<!--
1. Automatically close the issue when this PR is merged
    USAGE: Fixes #<issue number>
2. If your PR addresses an issue but does not close it
    USAGE: #<issue number> <reason>(i.e. This issue was worked on by @user and myself and his PR should close the issue.)
-->

## Type of Change

<!-- bug fix, feature, documentation, UI, etc. -->

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This PR adds GitHub stats for Telescope and Satellite.
Stats available: yearly/weekly number of commits, weekly lines added/removed and latest contributor.

<img width="1048" alt="Screen Shot 2021-11-13 at 6 02 04 PM" src="https://user-images.githubusercontent.com/53121061/141661564-fcc5515f-1805-4165-8aaf-65a68680adfe.png">


<!-- Please add a detailed description of what this PR does and why it is needed -->

## Checklist

<!-- Before submitting a PR, address each item -->

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)


<a href="https://gitpod.io/#https://github.com/Seneca-CDOT/telescope/pull/2466"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

menghif commented:
I haven't found a way to get the total number of contributors since Github only returns a max of 100:
https://docs.github.com/en/rest/reference/repos#list-repository-contributors
Any tips on how to get this?

humphd commented:
Request 1 item per page, and then look at the last page you're given for pagination in the `link` header they send in the response:

```
 curl -I -H "Accept: application/vnd.github.v3+json" "https://api.github.com/repos/Seneca-CDOT/telescope/contributors?per_page=1&anon=true"
HTTP/2 200
server: GitHub.com
date: Sat, 13 Nov 2021 23:55:18 GMT
content-type: application/json; charset=utf-8
...
link: <https://api.github.com/repositories/217872944/contributors?per_page=1&anon=true&page=2>; rel="next", <https://api.github.com/repositories/217872944/contributors?per_page=1&anon=true&page=153>; rel="last"
...
```

I think we have some other code that parses this, which you can steal.

Hey, I love what you're doing above!  Some initial thoughts

- Should we show something like "No commits this week" instead of "0" for commits, lines added, etc.  Let's find a way to deal with slow periods, of which there are many
- Can you make as many things as possible into links?  e.g., last contributor could link to the person.  We could also grab their avatar image and show it (https://github.com/{user}.png)
-

Andrewnt219 reviewed (commented):
This is awesome!

It may look better with all the cards having the same height. Suck that I don't have a better way to show you my changes, but here it is.
1. Add `height:100%` to the tag with class `card`
2. Add `justify-content: space-between` to the tag with class `card`
3. Add `margin-bottom: auto` to the tag with class `card-header`, or  `margin-top: auto` to the `hr`.

![image](https://user-images.githubusercontent.com/52666982/141663104-af1b3610-cccc-4039-b8ea-e1d76ce2f0ca.png)

Andrewnt219 commented on src/api/status/public/index.html:
```diff
@@ -32,6 +32,8 @@
     <!-- Font Awesome Icons -->
     <script src="https://kit.fontawesome.com/42d5adcbca.js" crossorigin="anonymous"></script>
+    <!-- GitHub Stats -->
+    <script type="text/javascript" src="gitHubStats.js"></script>
```
This needs to be moved before the closing `body` tag. The script file should also be inside the js folder.

https://github.com/Seneca-CDOT/telescope/blob/06c71038ad617c0ac805e4f3ccc9cbb1147862c3/src/api/status/public/pages/build.html#L32-L44

menghif replied:
done

menghif commented:
@humphd Thanks for the help on getting the total contributors!
I updated the "Last contributor" section and added the user's avatar. I also made the changes suggested by @Andrewnt219 😊

humphd commented on src/api/status/public/assets/js/gitHubStats.js:
```diff
@@ -0,0 +1,80 @@
+  let weeklyCommits = 0;
+
+  // get weekly commits for last year: https://docs.github.com/en/rest/reference/repos#get-the-weekly-commit-count
+  fetch(`https://api.github.com/repos/Seneca-CDOT/${repo}/stats/participation`, {
```
Write a function to build your GitHub API urls for you vs. hard-coding below:

```
const ghApiUrl = (repo = 'Seneca-CDOT', path) => `https://api.github.com/repos/Seneca-CDOT/${repo}{path}`
```

Then you can use it like `fetch(ghApiUrl('/stats/participation))`

humphd replied:
Reading further down, you can actually reuse a bunch of code here:

```js
const fetchGitHubApi = (repo = 'Seneca-CDOT', path) =>
  fetch(`https://api.github.com/repos/${repo}{path}`, {
    headers: {  Accept: 'application/vnd.github.v3+json', }
  })
  .then(res => {
    if(!res.ok) {
      throw new Error(`unable to get ${path}`);
    }
    return res.json()
  )
  .catch(err => console.error(err));
```

Now you can do:

```js
fetchGitHubApi('/stats/participation')
  .then(data => { /* process data here */);
```

menghif replied:
yes that's much better!

menghif replied:
I couldn't reuse some parts because I don't always return `res.json()`.

humphd commented on src/api/status/public/assets/js/gitHubStats.js:
```diff
@@ -0,0 +1,80 @@
+  fetchGitHubData('satellite');
+});
+
+response.headers.get('link');
```
What is this doing down here?  Where does `response` come from?

menghif replied:
oops not sure how that ended up there 😄

humphd commented on src/api/status/public/index.html:
```diff
@@ -1222,6 +1284,7 @@ <h6 class="mt-3">Thank you for sharing!</h6>
     <script src="assets/js/plugins/chartjs.min.js"></script>
     <script src="assets/js/demo-chart.js"></script>
     <script src="assets/js/demo-sidenav.js"></script>
+    <script src="assets/js/gitHubStats.js"></script>
```
Let's add a section specifically for our Telescope stuff

menghif replied:
inside `assets/js`? What should I call it?

humphd commented on src/api/status/src/server.js:
```diff
@@ -15,7 +15,12 @@ const satelliteOptions = {
+        connectSrc: [
+          "'self'",
+          '*.fontawesome.com',
+          `${process.env.API_HOST.replace(/(^\w+:|^)\/\//, '')}:4000`,
```
Was this fixed in another PR already?  cc @manekenpix

menghif commented on src/api/status/public/index.html:
```diff
@@ -351,61 +367,79 @@ <h4 class="mb-0">$53k</h4>
                 <p class="mb-0">
-                  <span class="text-success text-sm font-weight-bolder">+3% </span>than lask month
+                  Latest contributor:
+                  <a href="" title="" target="_blank" id="latest-commit-author-telescope">&nbsp;</a>
```
Do you mean adding `text-capitalize` to the name of the contributor? I think some people chose to have their GitHub name non capitalized.

Andrewnt219 commented on src/api/status/public/assets/css/material-dashboard.css:
```diff
@@ -5006,6 +5006,8 @@ fieldset:disabled .btn {
   position: relative;
   display: flex;
   flex-direction: column;
+  height: 100%;
```
I don't recommend overriding styles in the default asset, as it will affect the other components. You can add another stylesheet to override the default.

```css
/* homepage.css */
.card {
  height: 100%;
  justify-content: space-between
}
```

or make another class to make it clear (BEM style)
```css
/* homepage.css */
.card--homepage {
  height: 100%;
  justify-content: space-between
}
```

humphd replied:
Let's put `telescope-` in the name somehow, and maybe in a different directory, so it's clear that we are overriding the upstream styles.

Andrewnt219 replied:
Let's put it in `/public/css/homepage.css`, and import it after the default assets. 

For the class name, `.card--dashboard` should be clear about the override. Breakdown: "card" is the component, "dashboard"  is the modifier. The modifier indicates that it must be used with the "card" class and goes after the "card" class (i.e. className="card card--dashboard"). This follows the popular BEM methodology, which makes it easy for new contributors to follow.

humphd commented:
This needs a rebase too, to pick up the changes in server.js

Andrewnt219 commented on src/api/status/public/assets/js/github-stats.js:
```diff
@@ -0,0 +1,68 @@
+const fetchGitHubApi = (owner, repo, path) => {
```
Don't put it in `assets`, as these files are ignored by eslint and prettier. There should be an existing `public/js` directory.

menghif replied:
Thank you!

menghif commented:
I rebased and moved the js and css files based on the recommendations

Andrewnt219 reviewed (approved):
Awesome, looking forward to seeing it live.

humphd commented on src/api/status/public/js/github-stats.js:
```diff
@@ -0,0 +1,68 @@
+const fetchGitHubApi = (owner, repo, path) => {
```
If your arrow function immediately returns the result of an expression, you don't need a body or `return`:

```js
const fetchGitHubApi = (owner, repo, path) =>
  fetch(`https://api.github.com/repos/${owner}/${repo}/${path}`, {
    headers: { Accept: 'application/vnd.github.v3+json' },
  });
```

menghif replied:
fixed!

humphd reviewed (changes requested):
One small nit, otherwise this looks ready to go.
